### PR TITLE
[GenericCycle] Add a Cache for getExitBlocks in GenericCycle

### PR DESCRIPTION
**Pull request by Chengjunp**
In `UniformityAnalysis`, we need to get the exit blocks of cycles in the `DivergencePropagator` and currently, we have to do a search for the exit blocks every time. In this change, we add a cache of the results in the `GenericCycle` so that it can save the compile time. By testing, for some large cases, this can save about 60% compile time in the `UniformityAnalysis`.

### CODE REVIEW

**Comment by Chengjunp** [comment]
I’ve been investigating a compiletime issue and discovered that adding a cache for the `getExitBlocks` function in `GenericCycle` improves compile time of the `UniformityAnalysis` pass for some cases. However, I’m unsure whether this change aligns with the design principles of `GenericCycle`. If this change breaks some rules in the design, I also have another version that add the cache in the `UnformityAnalysis` pass. Since that is an analysis pass and the IR will not change during the process, I think that one would be safer than this change.

File: llvm/include/llvm/ADT/GenericCycleInfo.h [hunk 1]
```diff
@@ -74,12 +74,16 @@ template <typename ContextT> class GenericCycle {
   ///       always have the same depth.
   unsigned Depth = 0;
 
+  /// Cache for the results of GetExitBlocks
+  std::unique_ptr<SmallVector<BlockT *, 4>> ExitBlocksCache;
```
Comment (arsenm) [inline llvm/include/llvm/ADT/GenericCycleInfo.h:78]: Doesn't need to be unique_ptr

**Comment by ssahasra** [comment]
The cache seems useful, and I have always wondered why LoopInfo does not have something similar. But please note that all non-const member functions in GenericCycle and GenericCycleInfo should clear this cache. For example, `GenericCycle::setSingleEntry()` and `GenericCycleInfo::addBlockToCycle()`.

Adding @arsenm as reviewer to comment on whether having a cache like this is generally okay in LLVM. If we do put this cache, then GenericCycle is a better place than UniformityAnalysis, so that it gets wider use.

File: llvm/include/llvm/ADT/GenericCycleInfo.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (Chengjunp) [reply llvm/include/llvm/ADT/GenericCycleInfo.h:78]: The reason I declare the cache as a pointer is that the `getExitBlocks` is a const function, so I cannot modify it (update the cache) in that function. This is a WAR of it. I can also try to make this as a mutable vector.

Comment (Chengjunp) [reply llvm/include/llvm/ADT/GenericCycleInfo.h:78]: Same as https://github.com/llvm/llvm-project/pull/112290#discussion_r1802146978

[the remark this answers is not in the document]
Comment (arsenm) [reply llvm/include/llvm/ADT/GenericCycleInfo.h:78]: Yes, mutable is for this kind of case

File: llvm/include/llvm/ADT/GenericCycleInfo.h [hunk 2]
```diff
@@ -124,7 +141,12 @@ template <typename ContextT> class GenericCycle {
   bool contains(const GenericCycle *C) const;
 
   const GenericCycle *getParentCycle() const { return ParentCycle; }
-  GenericCycle *getParentCycle() { return ParentCycle; }
+
+  GenericCycle *getParentCycle() {
+    clearCache();
+    return ParentCycle;
+  }
```
Comment (arsenm) [inline llvm/include/llvm/ADT/GenericCycleInfo.h:148]: Clearing the cache in a getter (and only the non-const one) seems dubious

Comment (ssahasra) [reply llvm/include/llvm/ADT/GenericCycleInfo.h:148]: Yeah, clearly there is no change in state here and we don't need to clear the cache. At least this one is obvious enough. We could later revisit other non-const functions too, maybe as a future enhancement.

**Comment by Chengjunp** [comment]
Sorry for the late update. I just updated the code.

**Review by ssahasra — APPROVED** [review]
LGTM!

**Comment by Chengjunp** [comment]
Hi @arsenm , when you have a moment, could you please review the latest updates on this change? If everything looks good, would you be able to assist with merging it? Thanks very much for your help!

**Review by arsenm — APPROVED** [review]

**Pull request merged**